contracts/javascore/cluster-connection/src/main/java/xcall/adapter/cluster/ClusterConnection.java

```diff
@@ -0,0 +1,359 @@
+        if (response) {
+            BigInteger responseFee = responseFees.getOrDefault(to, BigInteger.ZERO);
+            return messageFee.add(responseFee);
+        }
+        return messageFee;
+    }
+
+    /**
+     * Sends a message to the specified network.
+     *
+     * @param to  Network Id of destination network
+     * @param svc name of the service
+     * @param sn  positive for two-way message, zero for one-way message, negative
+     *            for response(for xcall message)
+     * @param msg serialized bytes of Service Message
+     */
+    @Payable
+    @External
+    public void sendMessage(String to, String svc, BigInteger sn, byte[] msg) {
+        Context.require(Context.getCaller().equals(xCall.get()), "Only xCall can send messages");
+        BigInteger fee = BigInteger.ZERO;
+        if (sn.compareTo(BigInteger.ZERO) > 0) {
+            fee = getFee(to, true);
+        } else if (sn.equals(BigInteger.ZERO)) {
+            fee = getFee(to, false);
+        }
+
+        BigInteger nextConnSn = connSn.get().add(BigInteger.ONE);
+        connSn.set(nextConnSn);
+
+        Context.require(Context.getValue().compareTo(fee) >= 0, "Insufficient balance");
+        Message(to, nextConnSn, msg);
+    }
+
+    /**
+     * Receives a message from a source network.
+     *
+     * @param srcNetwork the source network id from which the message is received
+     * @param _connSn    the serial number of the connection message
+     * @param msg        serialized bytes of Service Message
+     * @param signatures array of signatures
+     */
+     @External
+     public void recvMessageWithSignatures(String srcNetwork, BigInteger _connSn, byte[] msg,
+                                           byte[][] signatures) {
+         OnlyRelayer();
+         Context.require(signatures.length >= validatorsThreshold.get().intValue(), "Not enough signatures");
+         byte[] messageHash = getMessageHash(srcNetwork, _connSn, msg);
+         List<Address> uniqueValidators = new ArrayList<>();
+         for (byte[] signature : signatures) {
+             Address validator = getValidator(messageHash, signature);
+             Context.require(isValidator(validator), "Invalid signature provided");
+             if (!uniqueValidators.contains(validator)) {
+                 uniqueValidators.add(validator);
+             }
+         }
+         Context.require(uniqueValidators.size() >= validatorsThreshold.get().intValue(), "Not enough valid signatures");
+         recvMessage(srcNetwork, _connSn, msg);
+     }
+
+    /**
+     * Receives a message from a source network.
+     *
+     * @param srcNetwork the source network id from which the message is received
+     * @param _connSn    the serial number of the connection message
+     * @param msg        serialized bytes of Service Message
+     */
+    @External
+    public void recvMessage(String srcNetwork, BigInteger _connSn, byte[] msg) {
```

Comment:
why do we have it here , wont this bypass signature validation?

Reply:
yeah, will remove this